Expose the scraping functions for usage in node.js

omarryhan opened the pull request.

omarryhan committed:
Exposed scraping functions

huchenme commented on src/index.js:
```diff
@@ -2,6 +2,7 @@ import axios from 'axios';
 import { sample, sampleSize, snakeCase } from 'lodash';
 import languages from './languages.json';
 import spokenLanguages from './spoken-languages.json';
+import * as scrapingFunctions from './functions/utils/fetch';
```
```suggestion
export {fetchDevelopers as scrapeDevelopers, fetchRepositories as scrapeRepositories} from './functions/utils/fetch';
```

omarryhan replied:
I thought that would work too, but it won't. Because in the `fetch.js` file, these two functions are exported with `module.exports = {...}` which is equivalent to `export default` and you cannot unpack an export default in an import statement (nor in an export statement).

huchenme reviewed (commented):
I dont see any issues here, but this will be cleaner

omarryhan committed:
Added a Typescript declaration file for index.js and bumped version from 0.0.0-development to 0.1.0

omarryhan commented:
I just added a Typescript declaration file as well.

omarryhan committed:
Reverted the change in version in package.json as this done independently

omarryhan commented:
I must say that the extra export I added will significantly increase the bundle size because it will ship Cheerio in the bundle, which is a bad idea because rarely does anyone scrape on the browser. I can remove the export, but then we won't be able to import it on server-side because right now Rollup isn't configured to do that. To solve that we can assign multiple outputs in `rollup.config.js` like shown [here](https://devhints.io/rollup). This would lets us remove the extra export I added in `index.js`.

After:

```js
  {
    input: 'src/index.js',
    external,
    output: [
      { file: pkg.main, format: 'cjs' },
      { file: pkg.module, format: 'es' },
    ],
    plugins: [
      json(),
      resolvePlugin,
      babelPlugin,
      process.env.NODE_ENV === 'production' && terser(),
    ],
  },
```

we'll add:

```js
  {
    input: 'src/functions/utils/fetch.js',
	output: {
		file: 'dist/scrape.js',
		format: 'cjs'
	},
	plugins: [
      resolvePlugin,
      babelPlugin,
    ],
  },
```

That way, if we want to import the scraping functions in Node.js, we do the following:

```js
import * as scrapingFunctions from '@huchenme/github-trending/scrape';

async () => {
	await scrapingFunctions.scrapeRepositories();
}()
```

What do you think?

huchenme commented on types/index.d.ts:
```diff
@@ -0,0 +1,85 @@
+declare module '@huchenme/github-trending' {
+    export interface FetchRepositoryOptions {
```
/nit can you make the indent 2 spaces instead?

huchenme commented on types/index.d.ts:
```diff
@@ -0,0 +1,85 @@
+
+    export const languages: Language[];
+    export const spokenLanguages: SpokenLanguage[];
+}
```
/nit newline here please

huchenme reviewed (commented):
thanks for keeping an eye on the bundle size, I think your approach will work, just some minor fix needed for stylings.

omarryhan committed:
Using exports. instead of module.exports = {  }
Fixed some incorrect types
Removed import of scraping code from index.js to not increase the bundle size on browsers
Seperated exported modules into index.js and scrape.js instead of just index.js
Added type declarations for scrape.js
Renamed dist to lib to be more expressive
Added lib to .eslintignore, .gitignore and .prettierignore
Added eslint to dev dependencies
Added new additions to README
Added eslint-plugin-jest
Disabled eslint-config-kentcdodds extensions because they were causing lots of unnecessary problems after I added eslint to the devDependencies

omarryhan commented:
I just separated both the scraping code and the code that calls the API so that the browser bundle doesn't grow.
I also addressed your comments about the code style. I added some other changes to accommodate the splitting of the code. Let me know if there's anything that's not clear.

omarryhan commented:
Right now, Rollup should be processing 4 files instead of just 1.

index.js
+ scrape.js
+ index.d.ts
+ scrape.d.ts

omarryhan committed:
Changed repo name to deploy on my own npm account

omarryhan committed:
Edit package.json info for new release

omarryhan committed:
Edit package.json info for new release

omarryhan committed:
edited version
